sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/ContainerRegistryAsyncClient.java

```diff
@@ -128,55 +91,71 @@
 
     Mono<PagedResponse<String>> listRepositoriesNextSinglePageAsync(String nextLink, Context context) {
         try {
-            Mono<PagedResponse<String>> pagedResponseMono = this.serviceClient.listRepositoriesNextSinglePageAsync(nextLink, context);
-            return pagedResponseMono.map(res -> getPagedResponseWithContinuationToken(res));
+            Mono<PagedResponse<String>> pagedResponseMono = this.registriesImplClient.getRepositoriesNextSinglePageAsync(nextLink, context);
+            return pagedResponseMono.map(res -> Utils.getPagedResponseWithContinuationToken(res));
         } catch (RuntimeException e) {
             return monoError(logger, e);
         }
     }
 
-
     /**
      * Delete the repository identified by `name`.
      *
      * @param name Name of the image (including the namespace).
      * @return deleted repository.
-     * @throws IllegalArgumentException thrown if parameters fail the validation.
-     * @throws AcrErrorsException thrown if the request is rejected by server.
-     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @throws ClientAuthenticationException thrown if the client's credentials do not have access to modify the namespace.
+     * @throws ResourceNotFoundException thrown if the given name was not found.
+     * @throws NullPointerException thrown if the name is null.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<DeleteRepositoryResult>> deleteRepositoryWithResponse(String name) {
```

Comment:
We should have tooling either as part of autogen and\or otherwise to build vanilla public builders - <AsyncClient> (public)XXX -> (package-private)XXXWithReponse(Context) -> (public)XXXWithResponse(Context) - SyncClient public(XXX) -> public XXXWithResponse(Context). All business logic can go into the Withresponse method and it will save a ton of time to have to hand author these.

<Not sure if we have an issue for this already - if not, will create one>

Reply:
Autorest generated different methods for XXX and XXXWithResponse and each end up calling the service directly instead of creating a waterfall - creating the need to test both branches. I personally think it will be more efficient to do it the other way round. Will talk to you or @jianghaolu about it.

---
In reply to: [609610749](https://github.com/Azure/azure-sdk-for-java/pull/20368#discussion_r609610749) [](ancestors = 609610749)